File: scripts/visualization.py
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import logging
import sys

# Set up logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

def ensure_directory(output_path):
    """Ensure the output directory exists."""
    os.makedirs(os.path.dirname(output_path), exist_ok=True)

def plot_host_microbe_ratio_pie(file_path, output_path):
    """Generate a pie chart for host vs microbe reads."""
    try:
        logging.info(f"Reading ratio file: {file_path}")
        
        # Check if file exists
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Ratio file not found: {file_path}")
        
        # Read and parse the ratio file
        with open(file_path, 'r') as file:
            lines = file.readlines()
            if len(lines) < 2:
                raise ValueError("Ratio file is not correctly formatted.")
            
            # Parse the host and microbial reads
            host_reads = int(lines[0].split(":")[1].strip())
            microbial_reads = int(lines[1].split(":")[1].strip())

        # Ensure the output directory exists
        ensure_directory(output_path)

        # Plot pie chart
        fig, ax = plt.subplots(figsize=(8, 8))
        labels = ['Host Reads', 'Microbial Reads']
        sizes = [host_reads, microbial_reads]
        ax.pie(sizes, labels=labels, autopct='%1.1f%%', startangle=140,
               colors=sns.color_palette("pastel"), pctdistance=0.85,
               wedgeprops=dict(width=0.3, edgecolor='w'))

        # Configure plot aesthetics
        ax.set_title('Host/Microbe Ratio', fontsize=16)
        ax.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.
        plt.tight_layout()

        # Save the pie chart
        plt.savefig(output_path)
        plt.close(fig)
        logging.info(f"Host/microbe ratio pie chart saved at {output_path}.")
    except Exception as e:
        logging.error(f"Error generating host/microbe ratio pie chart: {e}")
        raise

import textwrap

def plot_top_nonhost_bar(file_path, output_path):
    """Generate a vertical bar chart for the top non-host species with the sample name in the title."""
    try:
        logging.info(f"Reading top non-host file: {file_path}")
        
        # Check if file exists
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Top non-host file not found: {file_path}")
        
        # Read the top non-host CSV file
        df_microbes = pd.read_csv(file_path)
        
        # Extract the sample name from the column header (second column name)
        sample_name = df_microbes.columns[1]
        
        # Rename the columns for easier manipulation
        df_microbes.columns = ['Species', 'Count']
        
        # Log file content for debugging
        logging.info(f"Top microbes data:\n{df_microbes.head()}")

        # Ensure the output directory exists
        ensure_directory(output_path)

        # Wrap long species names
        df_microbes['Species'] = df_microbes['Species'].apply(lambda x: '\n'.join(textwrap.wrap(x, 20)))

        # Plot vertical bar chart for the top 10 microbes
        top_nonhost = df_microbes.head(10).sort_values(by="Count", ascending=False)
        fig, ax = plt.subplots(figsize=(12, 8))  # Increased figure size for better readability
        sns.barplot(x="Species", y="Count", data=top_nonhost, palette="viridis", ax=ax)

        # Add labels on bars
        ax.bar_label(ax.containers[0], fmt='%.0f', padding=3)

        # Configure plot aesthetics
        ax.set_title(f'Top 10 Microbe Species for Sample {sample_name}', fontsize=20)
        ax.set_xlabel('Species', fontsize=16)
        ax.set_ylabel('Count', fontsize=16)
        ax.tick_params(axis='x', labelsize=14, rotation=30)  # Adjusted rotation to 30 degrees for readability
        ax.tick_params(axis='y', labelsize=14)

        # Ensure tight layout and proper aspect ratio
        plt.tight_layout()

        # Save the bar chart
        plt.savefig(output_path, dpi=300)
        plt.close(fig)
        logging.info(f"Top 10 microbe species bar chart saved at {output_path}.")
    except Exception as e:
        logging.error(f"Error generating top 10 microbe species bar chart: {e}")
        raise


def main():
    try:
        # Check that we have received exactly 4 arguments
        if len(sys.argv) != 5:
            logging.error("Incorrect number of arguments provided. Expected 4 arguments.")
            sys.exit(1)

        # Assign command line arguments to variables
        top_nonhost_file = sys.argv[1]
        ratio_file = sys.argv[2]
        pie_chart_output = sys.argv[3]
        bar_chart_output = sys.argv[4]

        logging.info(f"Received arguments: {sys.argv}")

        # Generate visualizations
        plot_host_microbe_ratio_pie(ratio_file, pie_chart_output)
        plot_top_nonhost_bar(top_nonhost_file, bar_chart_output)

    except Exception as e:
        logging.error(f"Unexpected error occurred: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/visualization.r
library(ggplot2)
library(dplyr)
library(readr)

"""
This script generates visualizations, including:
1. A pie chart showing the ratio of host to microbe reads.
2. A bar chart displaying the top 10 non-host microbe species.

The script expects four command-line arguments:
    1. top_nonhost_file: Path to a CSV file containing the counts of non-host species.
    2. ratio_file: Path to a text file containing the ratio of host to microbe reads.
    3. pie_chart_output: Path to save the generated pie chart.
    4. bar_chart_output: Path to save the generated bar chart.
"""

args <- commandArgs(trailingOnly = TRUE)

if (length(args) != 4) {
  stop("Incorrect number of arguments provided. Expected 4 arguments: <top_nonhost_file> <ratio_file> <pie_chart_output> <bar_chart_output>")
}

top_nonhost_file <- args[1]
ratio_file <- args[2]
pie_chart_output <- args[3]
bar_chart_output <- args[4]

check_file_exists <- function(file_path) {
  if (!file.exists(file_path)) {
    stop(paste("File not found:", file_path))
  }
}

ensure_directory_exists <- function(output_path) {
  dir.create(dirname(output_path), showWarnings = FALSE, recursive = TRUE)
}

# Plot for host/microbe ratio pie chart
plot_host_microbe_ratio_pie <- function(file_path, output_path) {
  tryCatch({
    check_file_exists(file_path)
    
    lines <- readLines(file_path)
    if (length(lines) < 2) {
      stop("Ratio file is not correctly formatted. Something might be wrong.")
    }
    
    host_reads <- as.numeric(strsplit(lines[1], ":")[[1]][2])
    microbial_reads <- as.numeric(strsplit(lines[2], ":")[[1]][2])
    
    ensure_directory_exists(output_path)
    
    # Data for the pie chart
    pie_data <- data.frame(
      label = c("Host Reads", "Microbial Reads"),
      value = c(host_reads, microbial_reads)
    )
    
    # Plotting the pie chart
    pie_plot <- ggplot(pie_data, aes(x = "", y = value, fill = label)) +
      geom_bar(stat = "identity", width = 1, color = "white") +
      coord_polar("y") +
      theme_void() +
      theme(
        legend.title = element_blank(),
        legend.text = element_text(size = 16, color = "black"),
        plot.background = element_rect(fill = "white"),
        panel.background = element_rect(fill = "white"),
        plot.title = element_text(hjust = 0.5, size = 20, face = "bold", color = "#333333")
      ) +
      scale_fill_manual(values = c("#F8766D", "#00BFC4")) + # Modern contrasting colors
      geom_text(aes(label = paste0(round(value / sum(value) * 100), "%")),
                position = position_stack(vjust = 0.5), size = 6, color = "white", fontface = "bold") +
      labs(title = "Host vs Microbe Reads Distribution")
    
    ggsave(output_path, plot = pie_plot, width = 8, height = 8, dpi = 300)
    message(paste("Host/microbe ratio pie chart saved at", output_path))
    
  }, error = function(e) {
    stop(paste("Error generating host/microbe ratio pie chart:", e$message))
  })
}

# Plot for top non-host species bar chart
plot_top_nonhost_bar <- function(file_path, output_path) {
  tryCatch({
    check_file_exists(file_path)
    
    df_microbes <- read_csv(file_path)
    colnames(df_microbes) <- c("Species", "Count")
    
    ensure_directory_exists(output_path)
    
    # Select the top 10 species based on count
    top_nonhost <- df_microbes %>%
      top_n(10, Count) %>%
      arrange(desc(Count))
    
    # Plotting the bar chart with log scale for the y-axis
    bar_plot <- ggplot(top_nonhost, aes(x = reorder(Species, -Count), y = Count)) +
      geom_bar(stat = "identity", fill = "#00BFC4", color = "black", width = 0.7) +
      geom_text(aes(label = Count), vjust = -0.3, size = 4, color = "black") +
      scale_y_continuous(trans = 'log10', breaks = scales::trans_breaks("log10", function(x) 10^x)) +
      labs(title = "Top 10 Non-host Microbe Species", x = "Species", y = "Log(Count)") +
      theme_minimal(base_size = 16) +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1, size = 14, color = "black"),
        axis.text.y = element_text(size = 14, color = "black"),
        axis.title.x = element_text(size = 16, face = "bold", color = "#333333"),
        axis.title.y = element_text(size = 16, face = "bold", color = "#333333"),
        plot.title = element_text(size = 20, hjust = 0.5, face = "bold", color = "#333333"),
        panel.grid.major = element_line(color = "gray80"),
        panel.grid.minor = element_line(color = "gray90"),
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(fill = "white")
      )
    
    # Save the plot
    ggsave(output_path, plot = bar_plot, width = 12, height = 8, dpi = 300)
    message(paste("Top 10 microbe species bar chart saved at", output_path))
    
  }, error = function(e) {
    stop(paste("Error generating top 10 microbe species bar chart:", e$message))
  })
}

main <- function() {
  tryCatch({
    plot_host_microbe_ratio_pie(ratio_file, pie_chart_output)
    plot_top_nonhost_bar(top_nonhost_file, bar_chart_output)
    
  }, error = function(e) {
    stop(paste("Unexpected error occurred:", e$message))
  })
}

if (!interactive()) {
  main()
}
